judge/middleware.py

```diff
@@ -61,3 +64,32 @@ def __call__(self, request):
             request.in_contest = False
             request.participation = None
         return self.get_response(request)
+
+
+class APIMiddleware(object):
+    header_pattern = re.compile('^Bearer ([a-z2-7]{32})$', re.I)
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        full_token = request.META.get('HTTP_AUTHORIZATION', '')
+        if full_token:
+            token = self.header_pattern.match(full_token)
+            if token:
+                if request.path.startswith(reverse('admin:index')):
+                    return HttpResponse('Admin inaccessible', status=403)
+                try:
+                    request.user = User.objects.get(profile__api_token=token.group(1))
+                    request._cached_user = request.user
+                    request.csrf_processing_done = True
+                except User.DoesNotExist:
+                    response = HttpResponse('Invalid token')
+                    response['WWW-Authenticate'] = 'Bearer realm="API"'
+                    response.status_code = 401
+                    return response
+            else:
+                return HttpResponse('Invalid authorization header', status=400)
```

Comment:
Should this return 401 with WWW-Authenticate instead?

Reply:
Agreed upon 400